File: backend/src/tenants/tenants.controller.ts
import { BadRequestException, Body, Controller, Get, Param, Patch, Post, Req, UseGuards } from '@nestjs/common';
import { Request } from 'express';
import { CreateTenantDto } from './dto/create-tenant.dto';
import { UpdateTenantSelfDto } from './dto/update-tenant-self.dto';
import { UpdateTenantDto } from './dto/update-tenant.dto';
import { TenantsService } from './tenants.service';
import { AuthGuard } from '../common/guards/auth.guard';
import { AdminRoleGuard } from '../common/guards/admin-role.guard';

@Controller('tenants')
@UseGuards(AuthGuard)
export class TenantsController {
  constructor(private readonly tenantsService: TenantsService) {}

  @Get()
  list(@Req() request: Request & { auth?: any }) {
    const tenantId = request.auth?.role === 'tenant' ? request.auth?.tenantId : undefined;
    return this.tenantsService.list(tenantId);
  }

  @Post()
  @UseGuards(AdminRoleGuard)
  create(@Body() dto: CreateTenantDto) {
    return this.tenantsService.create(dto);
  }

  @Patch('me')
  updateSelf(@Req() request: Request & { auth?: any }, @Body() dto: UpdateTenantSelfDto) {
    const role = request.auth?.role;
    const tenantId = request.auth?.tenantId;
    if (role !== 'tenant' || !tenantId) {
      throw new BadRequestException('Tenant credentials required');
    }
    return this.tenantsService.updateSelf(tenantId, dto);
  }

  @Patch(':id')
  @UseGuards(AdminRoleGuard)
  update(@Param('id') id: string, @Body() dto: UpdateTenantDto) {
    return this.tenantsService.update(id, dto);
  }

  @Patch(':id/kill-switch')
  @UseGuards(AdminRoleGuard)
  toggleKillSwitch(@Param('id') id: string, @Body() body: { enabled: boolean }) {
    return this.tenantsService.toggleKillSwitch(id, body.enabled);
  }
}

File: backend-java/src/main/java/com/neria/manager/tenants/TenantsController.java
package com.neria.manager.tenants;

import com.neria.manager.common.entities.Tenant;
import com.neria.manager.common.security.AuthContext;
import jakarta.servlet.http.HttpServletRequest;
import java.util.List;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

@RestController
@RequestMapping("/tenants")
public class TenantsController {
  private final TenantsService tenantsService;

  public TenantsController(TenantsService tenantsService) {
    this.tenantsService = tenantsService;
  }

  @GetMapping
  public List<Tenant> list(HttpServletRequest request) {
    AuthContext auth = requireAuth(request);
    String tenantId = "tenant".equals(auth.getRole()) ? auth.getTenantId() : null;
    return tenantsService.list(tenantId);
  }

  @PostMapping
  public Tenant create(
      HttpServletRequest request, @RequestBody TenantsService.CreateTenantRequest dto) {
    requireAdmin(request);
    return tenantsService.create(dto);
  }

  @PatchMapping("/me")
  public Tenant updateSelf(
      HttpServletRequest request, @RequestBody TenantsService.UpdateTenantSelfRequest dto) {
    AuthContext auth = requireAuth(request);
    if (!"tenant".equals(auth.getRole()) || auth.getTenantId() == null) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Tenant credentials required");
    }
    return tenantsService.updateSelf(auth.getTenantId(), dto);
  }

  @PatchMapping("/{id}")
  public Tenant update(
      HttpServletRequest request,
      @PathVariable("id") String id,
      @RequestBody TenantsService.UpdateTenantRequest dto) {
    requireAdmin(request);
    return tenantsService.update(id, dto);
  }

  @PatchMapping("/{id}/kill-switch")
  public Tenant toggleKillSwitch(
      HttpServletRequest request,
      @PathVariable("id") String id,
      @RequestBody Map<String, Object> body) {
    requireAdmin(request);
    boolean enabled = Boolean.TRUE.equals(body.get("enabled"));
    return tenantsService.toggleKillSwitch(id, enabled);
  }

  private AuthContext requireAuth(HttpServletRequest request) {
    AuthContext auth = (AuthContext) request.getAttribute("auth");
    if (auth == null) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Unauthorized");
    }
    return auth;
  }

  private AuthContext requireAdmin(HttpServletRequest request) {
    AuthContext auth = requireAuth(request);
    if (auth.getRole() == null || !auth.getRole().equals("admin")) {
      throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Admin role required");
    }
    return auth;
  }
}
